aiida_raspa/workchains/base.py

```diff
@@ -45,18 +45,162 @@ def report_error_handled(self, calculation, action):
+            else:
+                conv_stat.append(False)
+
+        if not all(conv_stat):
+
+            self.report("Widom particle insertion calculationulation is NOT converged: repeating with more trials...")
+            self.ctx.inputs.retrieved_parent_folder = calculation.outputs['retrieved']
+            self.ctx.inputs.parameters = modify_number_of_cycles(self.ctx.inputs.parameters,
+                                                                 additional_init_cycle=Int(0),
+                                                                 additional_prod_cycle=Int(additional_cycle))
+            return ProcessHandlerReport(False)
+
+        return None
+
+    @process_handler(priority=410, enabled=False)
+    def check_gcmc_convergence(self, calc):
+        """Checks whether a GCMC calc is converged. Checking is based on the error bar on average loading."""
+        conv_threshold = 0.1
+        additional_init_cycle = 0
+        additional_prod_cycle = 0
+
+        output_gcmc = calc.outputs.output_parameters.get_dict()
+        structure_label = list(calc.get_incoming().nested()['framework'].keys())[0]
+        conv_stat = []
+
+        for comp in calc.inputs.parameters['Component']:
+
+            loading_average_comp = output_gcmc[structure_label]["components"][comp]["loading_absolute_average"]
+            loading_dev_comp = output_gcmc[structure_label]["components"][comp]["loading_absolute_dev"]
+
+            # It can happen for weekly adsorbed species.
+            # we need to think about a better way to handle it.
+            # Currently, if it happens for five iterations, self will not continue.
+            if loading_average_comp == 0:
+                conv_stat.append(False)
+            else:
+                error = round((loading_dev_comp / loading_average_comp), 2)
+                if error <= conv_threshold:
+                    conv_stat.append(True)
+                else:
+                    conv_stat.append(False)
+
+        if not all(conv_stat):
+            self.report("GCMC calculation is NOT converged: continuing from restart...")
+            self.ctx.inputs.retrieved_parent_folder = calc.outputs['retrieved']
+            self.ctx.inputs.parameters = modify_number_of_cycles(self.ctx.inputs.parameters,
+                                                                 additional_init_cycle=Int(additional_init_cycle),
+                                                                 additional_prod_cycle=Int(additional_prod_cycle))
+            return ProcessHandlerReport(False)
+
+        return None
+
+    @process_handler(priority=410, enabled=False)
+    def check_gemc_convergence(self, calc):
+        """Checks whether a GCMC calc is converged. Checking is based on the error bar on average loading which is
+        average number of molecules in each simulation box."""
+
+        conv_threshold = 0.1
+        additional_init_cycle = 0
+        additional_prod_cycle = 0
```

Comment:
Here, it needs to change to accept `conv_threshold` and `additional_init_cycle` and `additiona_prod_cycle` as arguments.